### ISSUE TITLE
[Navigation Animation] Allow observing of navigation transition state

### ISSUE THREAD

**Issue by ashar-7** [google/accompanist#633]
**Is your feature request related to a problem? Please describe.**
My root composable is a `Scaffold` which only shows top/bottom bars for the Bottom Navigation screens. Other screens have their own `Scaffold`s. The problem is, when I navigate with a slide-in animation from my bottom nav screens, the top/bottom bars of the root `Scaffold` are hidden before the slide in transition completes (because I'm observing the current screen in my root `Scaffold` and showing the top/bottom bars only if it's a bottom nav screen). This gives a bad UX and I'd like to keep the bars until the transition to the new screen completes.
```
// Common scaffold for Bottom Nav Screens
Scaffold(
    topBar = { if(currentScreen is BottomNavScreen) TopAppBar(...) },
    bottomBar = { if(currentScreen is BottomNavScreen) BottomNavigation(...) }
) {
    AnimatedNavHost() {
        // Bottom Nav Screens
        navigation(route = Main.route, startDestination = BottomNavScreen.Home.route) {
            composable(BottomNavScreen.Home.route) { HomeScreen() }
            composable(...) { ... }
            composable(...) { ... }
        }

        composable(Search.route, enterTransition = { slideInVertically() }, exitTransition = { slideOutVertically() }) {
            // Search Screen
            Scaffold(...) { ... }
        }
    }
}
```

**Describe the solution you'd like**
It'd be great if we can observe the nav transition state to hide the bars when the transition finishes, as suggested here https://kotlinlang.slack.com/archives/CJLTWPH7S/p1628271130295600?thread_ts=1628263708.267200&cid=CJLTWPH7S

**Comment by ianhanniballake** [comment]
Note that there is a pre-existing signal that the animations have completed - only once the animations complete does the `Lifecycle` of the current `NavBackStackEntry` move to `RESUMED` (it is purposefully capped at `STARTED` prior to the animations completing).

We can do more to make this an explicit signal though. What type of signal would you actually like? It seems like just having a `Boolean` signal of transition in progress wouldn't be quite enough for even your use case?

**Comment by ashar-7** [comment]
Something like this:
```
val isTransitioningToFullScreen = transitioningToFullScreen(transitionState.fromRoute, transitionState.toRoute)
...
if(currentScreen is BottomNavScreen || isTransitioningToFullScreen) { TopAppBar() }
```
where transitioningToFullScreen is a function I can define to check if we're transitioning from a BottomNavScreen to a screen where I want to hide the bars.
This is what comes to mind but not sure if this is the best way.

**Comment by ianhanniballake** [comment]
Apologies for not updating this issue sooner, but the API needed to handle this type of case was added in [Navigation 2.4.0-alpha10](https://developer.android.com/jetpack/androidx/releases/navigation#2.4.0-alpha10) via the new [`visibleEntries` property on `NavController`](https://developer.android.com/reference/kotlin/androidx/navigation/NavController#visibleEntries()).

This API provides a `StateFlow<List<NavBackStackEntry>>` that not only includes the current `NavBackStackEntry` (i.e., the one returned by `currentBackStackEntryAsState()`), but also any transitioning entries. As per the [docs](https://developer.android.com/reference/kotlin/androidx/navigation/NavController#visibleEntries()), the `currentState` of the `Lifecycle` of each `NavBackStackEntry` gives you additional information on exactly which entries are in the process of completing their exit transition (`CREATED`), running their enter transition (`STARTED`), or are the current topmost entry (`RESUMED`).

This `StateFlow` is updated whenever an entry is added or removed, which means any Composable that is using `collectAsState()` will automatically recompose whenever the list changes, thus allowing any logic you use to feed directly into `AnimatedVisibility` or similar APIs to update the visible state of other composables.

For example, let's say you have three destinations that should have a bottom bar (let's call them A, B, and C) and F, a full screen destination. This API would let you express logic like:
- If you're on A, B, or C or transitioning between them, keep showing the bottom bar
- If you navigate from A, B, or C to F, only hide the bottom bar after the transition to F completes

```
val visibleEntries by navController.visibleEntries.collectAsState()
// Show the bottom bar only while A, B, or C are still on the screen
val bottomNavDestinationVisible = visibleEntries.any { backStackEntry ->
  // Use any logic you want here
  backStackEntry.isBottomNavDestination()
}
AnimatedVisibility(visible = bottomNavDestinationVisible ) {
  // Your Bottom bar
}
```

This logic would immediately start showing the bottom bar as soon as you start the return transition from F back to A, B, or C (as starting that transition would cause A, B, or C to start appearing in `visibleEntries` again).

If you wanted the bottom bar to only start reappearing after you've settled back to A, B, or C, you could instead use logic such as:

```
val bottomNavDestinationVisible = visibleEntries.any { backStackEntry ->
  // Use any logic you want here
  backStackEntry.isBottomNavDestination()
}
val returningFromFullScreen = visibleEntries.any { backStackEntry ->
  backStackEntry.route == "F" && backStackEntry.lifecycle.currentState == Lifecycle.State.CREATED
}
val showBottomBar = bottomNavDestinationVisible && !returningFromFullScreen
AnimatedVisibility(visible = showBottomBar) {
  // Your Bottom bar
}
```

Let me know your thoughts on this API surface and if it fits your requirements.

**Comment by Nthily** [comment]
@ianhanniballake 
I‘m also interested in this topic, first I wanted to achieve an effect like this,

https://user-images.githubusercontent.com/31311826/140655014-42f97e97-48b3-441d-843e-e5afc20b8ff0.mp4

The `bottomBar` will move along with the new screen

I have tried what you described above

```
* If you're on A, B, or C or transitioning between them, keep showing the bottom bar
* If you navigate from A, B, or C to F, only hide the bottom bar after the transition to F completes
val visibleEntries by navController.visibleEntries.collectAsState()
// Show the bottom bar only while A, B, or C are still on the screen
val bottomNavDestinationVisible = visibleEntries.any { backStackEntry ->
  // Use any logic you want here
  backStackEntry.isBottomNavDestination()
}
AnimatedVisibility(visible = bottomNavDestinationVisible ) {
  // Your Bottom bar
}
```

But the result of this seems to be that the `bottomNavDestinationVisible` is updated only when the navigation transition animation of the new screen is finished.
In other words, the `bottomBar` will execute the `AnimatedVisibility` and disappear only after the new screen is fully displayed.


https://user-images.githubusercontent.com/31311826/140655267-4865ebf8-7d9a-4a3e-ae91-e9d7689c7188.mp4

Then i tried this method

```
val navBackStackEntry by navController.currentBackStackEntryAsState()
val currentDestination = navBackStackEntry?.destination?.route
```

I used the log to output the `currentDestination`, which outputs once before navigating to the new screen and once after navigating to the end, so I used it to determine if the `bottomBar` was displayed or not, and it did work, but I ran into some problems


https://user-images.githubusercontent.com/31311826/140655416-c1f1b166-2212-446a-af49-1a9adb43a206.mp4

My code looks like this:

```
            AnimatedNavHost(navController = navController, startDestination = AppScreen.bluetooth.route) {
                ....
                composable(AppScreen.settings.route,) {
                    Settings(navController)
                }
                composable(
                    route = "Test",
                    enterTransition = { initial, _ ->
                        when(initial.destination.route) {
                            AppScreen.settings.route -> slideInHorizontally(initialOffsetX = { 1000 }, animationSpec = tween(700))
                            else -> null
                        }
                    }
                ) {
                    Test()
                }
            }
```

```
@Composable
fun BottomBar(
    navController: NavController,
    pages: List<Screen>,
    isShown: Boolean
) {
    AnimatedVisibility(
        visible = isShown,
        exit = slideOutHorizontally(targetOffsetX = { -1000 }, animationSpec = tween(700))
    ) { ... }
```

I don't quite clear why there are extra white blocks in the lower left corner of the video, so I would like to know what I should do to achieve the effect shown in my video at the beginning.

And I have some other problems, if I use `AnimatedVisibility` in front of the bottomBar, then if I have many different Screens, I need to write a lot of judgments related to the navigation animation, otherwise the bottomBar may interfere with the navigation.

Also, if I want to achieve this transition effect like Twitter....

https://user-images.githubusercontent.com/31311826/140655749-0eb76f97-96b9-4f65-ac0f-8d0c41f3c7c2.mp4

As you can see, its new screen and the screen with the bottomBar are separated, if the NavHost is nested inside the Scaffold as I wrote before, it should be difficult to achieve this effect, right?

If I want to achieve this effect, what should I do? And, is it recommended to use nested NavHost?

**Comment by ianhanniballake** [comment]
@Nthily - your use case is a separate issue, perhaps more related to #729 in that you're trying to build a shared element that doesn't animate when you transition between certain screens but is treated as part of the content and its transition when moving to another screen.

This issue is specifically for "Allow observing of navigation transition state", which is what `visibleEntries` give you - a state change once a destination starts animating and again when it ends animating.
